# Тема “Обучение с учителем”


### *Дополнительные задания:
1). Загрузите датасет Wine из встроенных датасетов sklearn.datasets с помощью функции load_wine в переменную data.
2). Полученный датасет не является датафреймом. Это структура данных, имеющая ключи аналогично словарю. Просмотрите тип данных этой структуры данных и создайте список data_keys, содержащий ее ключи.
3). Просмотрите данные, описание и названия признаков в датасете. Описание нужно вывести в виде привычного, аккуратно оформленного текста, без обозначений переноса строки, но с самими переносами и т.д.4). Сколько классов содержит целевая переменная датасета? Выве
дите названия классов.
5). На основе данных датасета (они содержатся в двумерном массиве Numpy) и названий признаков создайте датафрейм под названием X.
6). Выясните размер датафрейма X и установите, имеются ли в нем пропущенные значения.
7). Добавьте в датафрейм поле с классами вин в виде чисел, имеющих тип данных numpy.int64. Название поля - 'target'.
8). Постройте матрицу корреляций для всех полей X. Дайте полученному датафрейму название X_corr.
9). Создайте список high_corr из признаков, корреляция которых с полем target по абсолютному значению превышает 0.5 (причем, само поле target не должно входить в этот список).
10). Удалите из датафрейма X поле с целевой переменной. Для всех признаков, названия которых содержатся в списке high_corr, вычислите квадрат их значений и добавьте в датафрейм X соответствующие поля с суффиксом '_2', добавленного к первоначальному названию признака. Итоговый датафрейм должен содержать все поля, которые, были в нем изначально, а также поля с признаками из списка high_corr, возведенными в квадрат. Выведите описание полей датафрейма X с помощью метода describe.


### Задание 1
Импортируйте библиотеки pandas и numpy.
Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [1]:
import pandas as pd
import numpy as np
import sklearn.linear_model as lr
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

In [3]:
lr = LinearRegression()

In [4]:
boston = load_boston()
X = pd.DataFrame(boston.data, columns=boston.feature_names)
y = pd.DataFrame(boston.target,columns=["target"])

In [5]:
X.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  506.000000  
mean    12.653063  
std      7.141062  
min      1.730000  
25%      6.950000  
50%     11.360000  
75%     16.955000  
max     37.970000

In [6]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


In [7]:
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   target  506 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB


In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=42)

In [9]:
X_train.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  354.000000  354.000000  354.000000  354.000000  354.000000  354.000000   
mean     3.469887   11.403955   11.133051    0.073446    0.557259    6.325672   
std      8.315831   22.608457    6.938651    0.261237    0.116792    0.719211   
min      0.009060    0.000000    1.210000    0.000000    0.385000    3.863000   
25%      0.082320    0.000000    5.130000    0.000000    0.450000    5.889500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.227500   
75%      3.281688   20.000000   18.100000    0.000000    0.631000    6.681750   
max     88.976200   95.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  354.000000  354.000000  354.000000  354.000000  354.000000  354.000000   
mean    68.799718    3.765874    9.437853  407.042373   18.277966  359.701808   
std     27.665360    2.126032    8.639971  166.522237    2.256792   86.924779   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     46.025000    2.073700    4.000000  279.000000   16.650000  376.782500   
50%     77.350000    3.107300    5.000000  330.000000   18.600000  391.600000   
75%     93.475000    5.400700   24.000000  666.000000   20.200000  396.172500   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  354.000000  
mean    12.421130  
std      7.112402  
min      1.730000  
25%      6.862500  
50%     10.925000  
75%     16.225000  
max     37.970000

In [10]:
X_test.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  152.000000  152.000000  152.000000  152.000000  152.000000  152.000000   
mean     3.948046   11.269737   11.145461    0.059211    0.548723    6.189059   
std      9.252844   24.983828    6.697032    0.236799    0.113877    0.654707   
min      0.006320    0.000000    0.460000    0.000000    0.392000    3.561000   
25%      0.079710    0.000000    5.692500    0.000000    0.448750    5.872750   
50%      0.243035    0.000000    9.690000    0.000000    0.532000    6.159500   
75%      4.367163    0.000000   18.100000    0.000000    0.614000    6.463500   
max     73.534100  100.000000   27.740000    1.000000    0.871000    8.725000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  152.000000  152.000000  152.000000  152.000000  152.000000  152.000000   
mean    68.051316    3.862975    9.809211  411.019737   18.869079  349.622500   
std     29.330844    2.062935    8.885388  173.666212    1.876845  100.693218   
min      6.200000    1.169100    1.000000  188.000000   13.000000    2.520000   
25%     39.700000    2.216750    4.000000  280.000000   17.800000  370.145000   
50%     78.500000    3.584950    5.000000  330.000000   19.200000  390.930000   
75%     95.000000    5.035675   24.000000  666.000000   20.200000  396.307500   
max    100.000000   10.710300   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  152.000000  
mean    13.193224  
std      7.201917  
min      2.880000  
25%      7.482500  
50%     12.020000  
75%     17.665000  
max     36.980000

In [11]:
# ?train_test_split

In [12]:
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [13]:
y_pred = lr.predict(X_test)

y_pred.shape

(152, 1)

In [14]:
check_test = pd.DataFrame({
    "y_test": y_test["target"],
    "y_pred": y_pred.flatten(),
})

check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

In [15]:
#?r2_score

In [16]:
r2_score(y_test,y_pred)

0.7112260057484948

### Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [17]:
model = RandomForestRegressor(max_depth=12, random_state=42)

In [18]:
model.fit(X_train,y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=12, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=100, n_jobs=None, oob_score=False,
                      random_state=42, verbose=0, warm_start=False)

In [19]:
y_rfr = model.predict(X_test)
y_rfr.shape

(152,)

In [20]:
check_test_rfr = pd.DataFrame({
    "y_test": y_test["target"],
    "y_pred": y_rfr.flatten(),
})

In [21]:
check_test_rfr.head(10)

y_test     y_pred
173    23.6  23.156214
274    32.4  31.199000
491    13.6  16.011514
72     22.8  24.227608
452    16.1  16.940542
76     20.0  21.847517
316    17.8  19.732158
140    14.0  14.911167
471    19.6  21.508747
500    16.8  21.239091

In [22]:
r2_score(y_test,y_rfr)

0.8691643189868505

## Выводы по сравнению моделей
На данном датасете лучше себя проявила модель RandomForestRegressor с результатом 0.8691643189868505 в модели LinearRegression  результат получился 0.7112260057484948.


### * Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [23]:
?RandomForestRegressor

In [25]:
sum(model.feature_importances_) Сумма всех показателей

1.0

In [31]:
model.feature_importances_

array([0.03177896, 0.00158242, 0.00646271, 0.0017163 , 0.01197879,
       0.43972879, 0.01160458, 0.06602488, 0.00504852, 0.01167048,
       0.02062718, 0.01151466, 0.38026172])

In [28]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [29]:
feature_weight = pd.DataFrame({
    "F_weight": model.feature_importances_,
    "F_name": boston.feature_names,
})

In [36]:
(feature_weight.sort_values(by=['F_weight'],ascending=False)).head(2)

F_weight F_name
5   0.439729     RM
12  0.380262  LSTAT

In [ ]:
### *Задание 4
В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection.Для этого датасета мы будем решать задачу классификации - будем определять,какие из транзакциции по кредитной карте являются мошенническими.Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки),так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель.Мы будем вычислять AUC, то есть площадь под кривой ROC.
Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.
Загрузите датасет creditcard.csv и создайте датафрейм df.
С помощью метода value_counts с аргументом normalize=True убедитесь в том, что выборка несбалансирована.Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков.Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма:
pd.options.display.max_columns = 100.
Просмотрите первые 10 строк датафрейма df.
Создайте датафрейм X из датафрейма df, исключив столбец Class.
Создайте объект Series под названием y из столбца Class.
Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: test_size=0.3, random_state=100, stratify=y.
У вас должны получиться объекты X_train, X_test, y_train и y_test.
Просмотрите информацию о их форме.
Для поиска по сетке параметров задайте такие параметры:
parameters = [{'n_estimators': [10, 15],
'max_features': np.arange(3, 5),
'max_depth': np.arange(4, 7)}]
Создайте модель GridSearchCV со следующими аргументами:
estimator=RandomForestClassifier(random_state=100),
param_grid=parameters,
scoring='roc_auc',
cv=3.
Обучите модель на тренировочном наборе данных (может занять несколько минут).
Просмотрите параметры лучшей модели с помощью атрибута best_params_.
Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.
Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.
Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументовмассивы y_test и y_pred_proba.